server/executor/transaction_runner.go

```diff
@@ -121,93 +117,81 @@ func (r persistentTransactionRunner) runTransaction(ctx context.Context, run mod
 	return r.updater.Update(ctx, run)
 }
 
-func (r persistentTransactionRunner) runTransactionStep(ctx context.Context, transactionRun model.TransactionRun, stepIndex int) (model.TransactionRun, error) {
-	step := transactionRun.Steps[stepIndex]
-	test, err := r.db.GetLatestTestVersion(ctx, step.ID)
+func (r persistentTransactionRunner) runTransactionStep(ctx context.Context, tr model.TransactionRun, step int, test model.Test) (model.TransactionRun, error) {
+	testRun := r.testRunner.Run(ctx, test, tr.Metadata, tr.Environment)
+	tr, err := r.updateStepRun(ctx, tr, step, testRun)
 	if err != nil {
-		return model.TransactionRun{}, fmt.Errorf("could not load transaction step: %w", err)
+		return model.TransactionRun{}, fmt.Errorf("could not update transaction run: %w", err)
 	}
 
-	testRun := r.testRunner.Run(ctx, test, transactionRun.Metadata, transactionRun.Environment)
-	stepRun := createStepRun(testRun)
-	transactionRun.StepRuns = append(transactionRun.StepRuns, stepRun)
-
+	done := make(chan bool)
 	// listen for updates and propagate them as if they were transaction updates
 	r.subscriptionManager.Subscribe(testRun.ResourceID(), subscription.NewSubscriberFunction(
 		func(m subscription.Message) error {
-			updatedRun := m.Content.(model.Run)
+			testRun := m.Content.(model.Run)
+			if testRun.LastError != nil {
+				tr.State = model.TransactionRunStateFailed
+				tr.LastError = testRun.LastError
+			}
 
-			for i, run := range transactionRun.StepRuns {
-				if run.ID == updatedRun.ID && run.TestID == updatedRun.TestID {
-					transactionRun.StepRuns[i] = createStepRun(updatedRun)
-					break
-				}
+			tr, err = r.updateStepRun(ctx, tr, step, testRun)
+			if err != nil {
+				done <- true
+				return err
 			}
 
 			r.subscriptionManager.PublishUpdate(subscription.Message{
-				ResourceID: transactionRun.ResourceID(),
+				ResourceID: tr.ResourceID(),
 				Type:       "result_update",
-				Content:    transactionRun,
+				Content:    tr,
 			})
 
-			return nil
-		}),
-	)
-
-	err = r.updater.Update(ctx, transactionRun)
-	if err != nil {
-		return model.TransactionRun{}, fmt.Errorf("could not update transaction run: %w", err)
-	}
-
-	var wg sync.WaitGroup
-	wgPtr := &wg
-	wg.Add(1)
-
-	r.subscriptionManager.Subscribe(testRun.TransactionStepResourceID(), subscription.NewSubscriberFunction(
-		func(m subscription.Message) error {
-			defer wgPtr.Done()
-			runResult := m.Content.(RunResult)
+			fmt.Println("ACA", testRun.State, testRun.State.IsFinal())
+			if testRun.State.IsFinal() {
+				done <- true
+			}
 
-			return r.updateTransactionRun(ctx, stepIndex, &transactionRun, runResult)
+			return nil
 		}),
 	)
+	// TODO: this will block indefinitely. we need to set a timeout or something
+	<-done
```

Comment:
Without the timeout, this can potentially break tracetest again. You probably can use a simple select here. Probably something like:

```suggestion
   select {
     case <-done:
       return tr, err
     case <- time.After(timeout):
       return tr, fmt.Errorf("transaction step execution timed out")
   }
```

Reply:
I thought of this, but how much time is a timeout? It will depend on the time it takes for the test to finish. Tests don't have a timeout other than the http and trace polling, so any timeout we put here will be the same as waiting for the test to finish.

This comment is not meant for now, but for later, when we implement a more advanced logic for timing tests etc